docs/config.json.md
# config.json explanation

## Configuration free
- `config.json` file is optional, you can run a `did` node without a `config.json` file.

## Change active network
Just modify the `ActiveNet` parameter in the `config.json` file.
```json
{
  "ActiveNet": "test"
}
```
Default config for `testnet`
- Peer-to-Peer network connect to ELA `testnet`.

## Inline Explanation

```json5
{
  "ActiveNet": "mainnet", // Network type. Choices: mainnet testnet and regnet
  "Magic": 2019004,       // Magic number：Segregation for different subnet. No matter the port number, as long as the magic number not matching, nodes cannot talk to each others
  "NodePort": 20618,      // The port number for P2P network connection.
  "DisableDNS": false,    // Disable the DNS service of the P2P network.
  "PermanentPeers": [     // Specify a list of peers to connect with permanently.
    "localhost:20618"
  ],
  "SPVMagic": 2017001,    // SPV magic number.
  "SPVDisableDNS": false, // Disable the DNS service of the SPV P2P network.
  "SPVPermanentPeers": [  // Specify a list of SPV peers to connect with permanently.
    "localhost:20338"
  ],
  "CRCArbiters": [        // CRCArbiters is the list of CRC public key.
    "03e435ccd6073813917c2d841a0815d21301ec3286bc1412bb5b099178c68a10b6",
    "038a1829b4b2bee784a99bebabbfecfec53f33dadeeeff21b460f8b4fc7c2ca771",
    "02435df9a4728e6250283cfa8215f16b48948d71936c4600b3a5b1c6fde70503ae",
    "027d44ee7e7a6c6ff13a130d15b18c75a3b47494c3e54fcffe5f4b10e225351e09",
  ],
  "ExchangeRate": 1.0,    // Defines the exchange rate of main/side asset.
  "MinCrossChainTxFee": 10000, // Defines the minimum fee for a cross chain transaction.
  "EnableREST": false,    // Enable the RESTful service.
  "RESTPort": 20604,      // Specify a port for the RESTful service.
  "EnableWS": false,      // Enable the WebSocket service.
  "WSPort": 20605,        // Specify a port for the WebSocket service.
  "EnableRPC": false,     // Enable the JSON-RPC service.
  "RPCPort": 20606,       // Specify a port for the JSON-RPC service.
  "RPCUser": "User",      // Specify the username when accessing the JSON-RPC service.
  "RPCPass": "Password",  // Specify the password when accessing the JSON-RPC service.
  "RPCWhiteList": [       // Specify a white list of IP addresses for the JSON-RPC service.
    "127.0.0.1"
  ],
  "LogLevel": 1,          // Specify the debug log level. Choices: (0)debug (1)info (2)warn (3)error (4)fatal (5)off.
  "PerLogFileSize": 20,   // Specify the maximum file size of a single log file.
  "LogsFolderSize": 2048, // Specify the maximum logs folder size which containing all log files.
  "DisableTxFilters": false, // Disable the transaction filer function.
  "EnableMining": false,  // Enable the mining service which can produce blocks.
  "InstantBlock": false,  // Set the block producing to instant mode which can make the mining service produce block instantly.
  "PayToAddr": "8VYXVxKKSAxkmRrfmGpQR2Kc66XhG6m3ta", // Specify the account address to receive rewards by mining blocks.
  "MinerInfo": "ELA",     // The miner info displaying the miner in coinbase transaction.
}
```


params/params.go
package params

import (
	"math/big"
	"time"

	"github.com/elastos/Elastos.ELA.SideChain/auxpow"
	"github.com/elastos/Elastos.ELA.SideChain/config"
	"github.com/elastos/Elastos.ELA.SideChain/types"
	"github.com/elastos/Elastos.ELA/common"
	ela "github.com/elastos/Elastos.ELA/core/types"
	"github.com/elastos/Elastos.ELA/core/types/payload"
)

// These variables are the chain consensus parameters.
var (
	// elaAsset is the transaction that create and register the ELA coin.
	elaAsset = types.Transaction{
		TxType:         types.RegisterAsset,
		PayloadVersion: 0,
		Payload: &types.PayloadRegisterAsset{
			Asset: types.Asset{
				Name:      "ELA",
				Precision: 0x08,
				AssetType: 0x00,
			},
			Amount:     0 * 100000000,
			Controller: common.Uint168{},
		},
		Attributes: []*types.Attribute{},
		Inputs:     []*types.Input{},
		Outputs:    []*types.Output{},
		Programs:   []*types.Program{},
	}

	// genesisTime indicates the time when ELA genesis block created.
	genesisTime, _ = time.Parse(time.RFC3339, "2019-03-05T00:00:00Z")

	// genesisHeader
	genesisHeader = types.Header{
		Version:    types.BlockVersion,
		Previous:   common.Uint256{},
		MerkleRoot: ElaAssetId,
		Timestamp:  uint32(genesisTime.Unix()),
		Bits:       0x1d03ffff,
		Nonce:      types.GenesisNonce,
		Height:     uint32(0),
		SideAuxPow: auxpow.SideAuxPow{
			SideAuxBlockTx: ela.Transaction{
				TxType:         ela.SideChainPow,
				PayloadVersion: payload.SideChainPowVersion,
				Payload:        new(payload.SideChainPow),
			},
		},
	}

	// bigOne is 1 represented as a big.Int.  It is defined here to avoid
	// the overhead of creating it multiple times.
	bigOne = big.NewInt(1)

	// powLimit is the highest proof of work value a block can have for the network.
	//  It is the value 2^255 - 1.
	powLimit = new(big.Int).Sub(new(big.Int).Lsh(bigOne, 255), bigOne)

	// "8VYXVxKKSAxkmRrfmGpQR2Kc66XhG6m3ta"
	mainNetFoundation = common.Uint168{
		0x12, 0x9e, 0x9c, 0xf1, 0xc5, 0xf3, 0x36,
		0xfc, 0xf3, 0xa6, 0xc9, 0x54, 0x44, 0x4e,
		0xd4, 0x82, 0xc5, 0xd9, 0x16, 0xe5, 0x06,
	}

	// "8NRxtbMKScEWzW8gmPDGUZ8LSzm688nkZZ"
	testNetFoundation = common.Uint168{
		0x12, 0x50, 0x96, 0x58, 0xd3, 0x9e, 0x4b,
		0xde, 0x30, 0x79, 0xe3, 0xf8, 0xde, 0x91,
		0xf4, 0x9c, 0xaa, 0x97, 0x01, 0x5c, 0x9e,
	}

	// ELAAssetID represents the asset ID of ELA coin.
	ElaAssetId = elaAsset.Hash()

	// GenesisBlock represent the genesis block of the Token chain.
	GenesisBlock = &types.Block{
		Header:       genesisHeader,
		Transactions: []*types.Transaction{&elaAsset},
	}
)

// MainNetParams defines the network parameters for the main network.
var MainNetParams = config.Params{
	Name:        "mainnet",
	Magic:       2019004,
	DefaultPort: 20618,

	DNSSeeds: []string{
		"node-mainnet-005.elastos.org:20618",
		"node-mainnet-010.elastos.org:20618",
		"node-mainnet-015.elastos.org:20618",
		"node-mainnet-020.elastos.org:20618",
		"node-mainnet-025.elastos.org:20618",
	},

	Foundation:           mainNetFoundation,
	ElaAssetId:           ElaAssetId,
	GenesisBlock:         GenesisBlock,
	PowLimit:             powLimit,
	PowLimitBits:         0x1f0008ff,
	TargetTimespan:       24 * time.Hour,  // 24 hours
	TargetTimePerBlock:   2 * time.Minute, // 2 minute
	AdjustmentFactor:     4,               // 25% less, 400% more
	CoinbaseMaturity:     100,
	MinTransactionFee:    100,
	ExchangeRate:         1,
	MinCrossChainTxFee:   10000,
	CheckPowHeaderHeight: 31538,
}

// TestNetParams defines the network parameters for the test network.
var TestNetParams = testNetParams(MainNetParams)

// RegNetParams defines the network parameters for the regression network.
var RegNetParams = regNetParams(MainNetParams)

// testNetParams returns the network parameters for the test network.
func testNetParams(cfg config.Params) config.Params {
	cfg.Name = "testnet"
	cfg.Magic = 2019104
	cfg.DefaultPort = 21618
	cfg.DNSSeeds = []string{
		"node-testnet-002.elastos.org:21618",
		"node-testnet-003.elastos.org:21618",
		"node-testnet-004.elastos.org:21618",
	}
	cfg.Foundation = testNetFoundation
	cfg.CheckPowHeaderHeight = 30000
	return cfg
}

// regNetParams returns the network parameters for the regression network.
func regNetParams(cfg config.Params) config.Params {
	cfg.Name = "regnet"
	cfg.Magic = 2019204
	cfg.DefaultPort = 22618
	cfg.DNSSeeds = []string{
		"node-regtest-102.eadd.co:22618",
		"node-regtest-103.eadd.co:22618",
		"node-regtest-104.eadd.co:22618",
	}
	cfg.Foundation = testNetFoundation
	cfg.CheckPowHeaderHeight = 20000
	return cfg
}

// InstantBlock changes the given network parameter to instant block mode.
func InstantBlock(cfg *config.Params) {
	cfg.PowLimitBits = 0x207fffff
	cfg.TargetTimespan = 1 * time.Second * 10
	cfg.TargetTimePerBlock = 1 * time.Second
}
